components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Runtime/Extensions/OpcConfigEx.cs

```diff
@@ -40,38 +41,58 @@ public static class OpcConfigEx {
+            applicationConfiguration.CertificateValidator
+                .Update(applicationConfiguration.SecurityConfiguration).ConfigureAwait(false);
 
-            // use existing certificate, if it is there
-            certificate = applicationConfiguration.SecurityConfiguration.ApplicationCertificate.Find(true).Result;
+            // use existing certificate, if present
+            var certificate = applicationConfiguration.SecurityConfiguration
+                .ApplicationCertificate.Certificate;
 
             // create a self signed certificate if there is none
             if (certificate == null && createSelfSignedCertIfNone) {
                 certificate = CertificateFactory.CreateCertificate(
-                    applicationConfiguration.SecurityConfiguration.ApplicationCertificate.StoreType,
-                    applicationConfiguration.SecurityConfiguration.ApplicationCertificate.StorePath,
+                    applicationConfiguration.SecurityConfiguration
+                        .ApplicationCertificate.StoreType,
+                    applicationConfiguration.SecurityConfiguration
+                        .ApplicationCertificate.StorePath,
                     null,
                     applicationConfiguration.ApplicationUri,
                     applicationConfiguration.ApplicationName,
-                    applicationConfiguration.ApplicationName,
+                    configuredSubject,
                     null,
                     CertificateFactory.defaultKeySize,
                     DateTime.UtcNow - TimeSpan.FromDays(1),
                     CertificateFactory.defaultLifeTime,
                     CertificateFactory.defaultHashSize
                 );
 
-                // update security information
-                applicationConfiguration.SecurityConfiguration.ApplicationCertificate.Certificate =
-                    certificate ??
+                if (certificate == null) {
                     throw new Exception(
                         "OPC UA application certificate can not be created! Cannot continue without it!");
-                //await applicationConfiguration.CertificateValidator.UpdateCertificate(
-                   //applicationConfiguration.SecurityConfiguration).ConfigureAwait(false);
+                }
+
+                applicationConfiguration.SecurityConfiguration
+                    .ApplicationCertificate.Certificate = certificate;
+                try {
+                    // copy the certificate *public key only* into the trusted certificates list
+                    using (ICertificateStore trustedStore = applicationConfiguration
+                        .SecurityConfiguration.TrustedPeerCertificates.OpenStore()) {
+                        using (var publicKey = new X509Certificate2(certificate.RawData)) {
+                            trustedStore.Add(publicKey.YieldReturn());
+                        }
+                    }
+                }
+                catch { }
```

Comment:
this is not needed, or I'd not recommend it. only if you need to trust yourself doing client server on the same host (similar to autoaccept)

Reply:
I think we need to live with such compromises until we properly integrate the vault respective signed certificates. We had once an ask to get rid of the AutoAccept=true flag on Connected Factory and this was the solution since all modules and servers were running on the same host and were configured to basically use the same trusted certs store. Another reason is the sporadic test failures during build...